react/features/local-recording/recording/WavAdapter.js

```diff
@@ -0,0 +1,284 @@
+import { RecordingAdapter } from './RecordingAdapter';
+import { downloadBlob, timestampString } from './Utils';
+
+const logger = require('jitsi-meet-logger').getLogger(__filename);
+
+const WAV_BITS_PER_SAMPLE = 16;
+const WAV_SAMPLE_RATE = 44100;
+
+/**
+ * Recording adapter for raw WAVE format.
+ */
+export class WavAdapter extends RecordingAdapter {
+
+    _audioContext = null;
+    _audioProcessingNode = null;
+    _audioSource = null;
+
+    _wavLength = 0;
+    _wavBuffers = [];
+    _isInitialized = false;
+
+    /**
+     * Constructor.
+     *
+     */
+    constructor() {
+        super();
+
+        this._saveWavPCM = this._saveWavPCM.bind(this);
+    }
+
+    /**
+     * Implements {@link RecordingAdapter#ensureInitialized()}.
+     *
+     * @inheritdoc
+     */
+    ensureInitialized() {
+        if (this._isInitialized) {
+            return Promise.resolve();
+        }
+
+        const p = new Promise((resolve, reject) => {
+            navigator.getUserMedia(
+
+                // constraints - only audio needed for this app
+                {
+                    audioBitsPerSecond: WAV_SAMPLE_RATE * WAV_BITS_PER_SAMPLE,
+                    audio: true,
+                    mimeType: 'application/ogg' // useless?
+                },
+
+                // Success callback
+                stream => {
+                    this._audioContext = new AudioContext();
+                    this._audioSource
+                     = this._audioContext.createMediaStreamSource(stream);
+                    this._audioProcessingNode
+                      = this._audioContext.createScriptProcessor(4096, 1, 1);
+                    this._audioProcessingNode.onaudioprocess = e => {
+                        const channelLeft = e.inputBuffer.getChannelData(0);
+
+                        // https://developer.mozilla.org/en-US/docs/
+                        // Web/API/AudioBuffer/getChannelData
+                        // the returned value is an Float32Array
+                        this._saveWavPCM(channelLeft);
+                    };
+                    this._isInitialized = true;
+                    resolve();
+                },
+
+                // Error callback
+                err => {
+                    logger.error(`Error calling getUserMedia(): ${err}`);
+                    reject();
+                }
+            );
+        });
+
+        return p;
+    }
+
+    /**
+     * Implements {@link RecordingAdapter#start()}.
+     *
+     * @inheritdoc
+     */
+    start() {
+        return new Promise(
```

Comment:
Should start use ensureInitialized before doing the starting logic?

Reply:
Hmm. The only caller of this method is `RecordingController`, and `ensureInitialized` is chained there (https://github.com/ztl8702/jitsi-meet/blob/87e9d4164fa7c4222acf7490f0ea55d572e15c46/react/features/local-recording/controller/RecordingController.js#L401)  Maybe the interface can be simplified in this case.